### PR TITLE
Add Tinker Complement's High Oven support

### PR DESCRIPTION
**Pull request by Riernar**
Hi @jaredlll08 ,

I added @KnightMiner 's tcomplement support to ModTweaker, as asked for in #718 . I tried keeping the same code style as other tcomplement support and used a new package to organise code.

@KnightMiner , here's the PR. I hope I'm using your API correctly - let me know if anything is amiss !

### CODE REVIEW

**Comment by Riernar** [comment]
For testing purposes, here is a small script that uses just CraftTweaker, ModTweaker, Mantle-TinkerConstruct-Tcomplement.
It requires a tcomplement recompiled from source though, as a change was needed in the source code to implement the compatibility.

```
import mods.tcomplement.highoven.HighOven;
import mods.tcomplement.highoven.IMixRecipe;

HighOven.removeFuel(<minecraft:coal:1>);
HighOven.addFuel(<minecraft:diamond>, 1000, 25);

HighOven.removeHeatRecipe(<liquid:steam>);
HighOven.addHeatRecipe(<liquid:lava> * 100, <liquid:water> * 1000, 450);

HighOven.removeMixRecipe(<liquid:steel>);
HighOven.newMixRecipe(<liquid:cobalt> * 144, <liquid:iron> * 144, 1000)
	.addOxidizer(<minecraft:stone>, 25)
	.addReducer(<minecraft:cobblestone>, 35)
	.addPurifier(<minecraft:redstone>, 45)
	.addPurifier(<ore:stickWood>, 100)
	.register();
```

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Don't add 300 here or its inconsistent with the melter and smeltery recipes which use Kelvin in ModTweaker. There is a Tinkers Construct config option to show all temperatures in Kelvin so they match the ModTweaker recipes for pack makers, which is why its different from whats shown in the GUI/JEI.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: Same to above about temperatures, everything else in crafttweaker uses Kelvin so don't add 300 here for consistency

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: I don't like special casing this to ItemStack and oredict as that limits the options, in my overall review comment I give a suggestion of how to handle this.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: Same as above, instead of special casing this to IItemStack and IOreDictEntry, I'd suggest using the method I describe in the main review comment to handle any IIngredient type.

I will point out if you use RecipeMatch directly, set  `amountMatched` to the `consumeChance` variable in the constructor as that is how its set internally. Might even let you replace the map with just a list if you convert to RecipeMatch in the add methods

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Just use `event.getRecipe().matches(FluidStack,FluidStack)`, I coded it originally for this purpose. The method takes two fluid stack parameters treating null as a wildcard for each, and can be overridden if someone adds a custom high oven recipe type (so your code will still match properly)

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Diddo to above, use `event.getRecipe().matches(FluidStack,FluidStack)`. This also should probably check the input as an optional parameter for removing, in case there are multiple ways to get steel for example

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: This comes down to preference, but if you wanted to you could have just one list of additives and use `MixAdditive` to distinguish. Then when iterating the list you can just use the generic `recipe.addAdditive(MixAdditive,RecipeMatch)` method.

**Review by KnightMiner — CHANGES_REQUESTED** [review]
This looks good overall, good job dealing with the complexities of `IMixRecipe` using the helper, I was thinking something similar to that would be needed to handle the additives.

One of my main comments is how you handle converting `IIngredient` to `RecipeMatch`. Your current method will handle item stacks and oredict names, but won't work for more special ingredient types such as NBT ingredients (recipe match has to be told to use NBT) or crafttweakers little known fluid container ingredient (ILiquidStack when used in an item context matches fluid containers, recipe match does not have anything like that at all). I made a utility in [my PR to update Inspirations](https://github.com/jaredlll08/ModTweaker/pull/729/files#diff-3530a6add6edb64334eaaf9029a2ca22) which handles this by just wrapping the `IIngredient` in a special `RecipeMatch` which redirects the matches method. Feel free to copy over to your PR (we will just need to deal with the merge conflict based on which one Jared merges first). Note you will need the constructor with `amountMatched` for the mix recipe additives to pass in the consume chance parameter (as that's where its stored internally)

Another note is there is no method to add or remove additives for an existing mix recipe. The Tinkers Complement API does support it, you could use the following design to accomplish it:
* Start with a `HighOven` method called something like `findMixRecipe`
    * This would behave very similarly to `newMixRecipe` except it looks for an existing recipe rather than making a new one.
    * You could make it match similarly to the remove methods, with input and output both supporting null as wildcard.
* The class returned would have methods such as `addReducer`, etc. which populate a list to add, and methods like `removeReducer`, etc. which populate a list to remove.
    * Skip any of the getters for input, output, or the additives, we have no way of knowing those until after the additives are already added
* To add additives, listen to `HighOvenMixRegisterEvent`. Use `event.getRecipe().matches(FluidStack,FluidStack)` to determine if its the right recipe then iterate the list(s) of additives to add
* To remove additives, you can listen to `HighOvenMixAdditiveEvent`. If `event.getRecipe().matches(FluidStack,FluidStack)`, `event.getType()` is the right additive type, and `event.getAdditive()` matches an ingredient in the list, cancel the event

If you would rather leave this out for now, I could make a follow up PR, though it should be pretty straight forward if you start with the way you did `newMixRecipe`. As a side note, sorry for such a complicated recipe structure.

Anyways, overall it looks pretty good, thanks for doing this (saves me a lot of time writing the crafttweaker support).

**Comment by KnightMiner** [comment]
You should not need to compile from the source for this to work, build `0.4.2.39` has the change from your PR in it which can be pulled from Maven

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Oh, that's easier and safer, i.e. definitely better - I'll use that.

I realized their was a problem with removing heat/mix recipes for a given output, but for multiple inputs (e.g. remove **iron -> steel** and **lava -> steel** but not **aluminium -> steel**). Due to the `Map<>` structure, only the last one would be applied.
I switched to a `Map<K, Set<V> >` structure but I'm unsure how to handle the recipe comparison. I see two options:
+ The keys are `K instanceof FluidStack`, conversion from `ILiquidStack` happens in the `remove` method. When cancelling events, I `get(event.getRecipe().getOutput())` the `Set<FluidStack>` of forbidden inputs from the `Map`, and then check if `event.getRecipe().getInput()` is in the set. This prevents iterating on all removed entries for each event, **but** I'm not sure the ILiquidStack comparison will be correct and handle all cases
+ (current) I do not use the `get` method and blindly iterate on all entries in the `Map<ILiquidStack, Set<ILiquidStack> >`, checking each time if `event.getRecipe().matches()` holds. I'm sure the comparison will be correct, but the iteration is sub-optimal.

In both case, a mapping from `K` to `null` indicate the method with null input as wildcard was used, and I cancel all event with the given output.

Should I switch to map/set lookup ? The performance gain should be small and I'm not sure it'll work properly, so it might not be worth the change.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: I should have looked at the other support instead of just putting everything in °C. Corrected

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: Same as above, corrected

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: I shamelessly copy-pasted your `RecipeMatchIIngredient`. This is way cleaner and versatile, thanks !

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: Switched other to `RecipeMatchIIngredient` for additives. I kept maps in MixRecipeHelper for easier additives adding/remove/change, but `AddMixRecipe` just uses a `List<SimpleEntry<>>` as it is essentially immutable.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Added input to the check, otherwise same as above for check considerations.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeHelper.java]: `AddMixRecipe` now does that (but not `MixRecipeHelper`)

**Comment by Riernar** [comment]
Hi @KnightMiner ,

Happy to help ! I needed that for my own pack, so its not completely selfless.
Your code interface is great, it easy to hook into it and add/remove stuff. And from a pack making point of view, your recipes are very versatile and offers a tons of possibilities for e.g. metallurgy, which is awesome ! No worries about the interface :)

I integrated your corrections, shamelessly copy-pasting your `IIngredient` `RecipeMatch` wrapper, and added a class for manipulating already existing mix recipes (including scripted ones).

There is now a `MixRecipeBuilder` (the old `MixRecipeHelper` / `mods.tcomplement.highoven.IMixRecipe`) and a `MixRecipeManager`. The interface of the former is as before, while the manager allows to edit other mix recipes, including scripted ones.

I'm not sure I'm checking the matching of `RecipeMatch` for additive removal in the right order, could you confirm that ? I want it to cancel the event if the added additive matches the `IIngredient` in any way. `null` can be used to remove all additives of a type.

Anyway, thanks a lot for your help, the code is much cleaner and safer now !

PS. I should really learn a bit about maven and gradle at some point, thanks for pointing out jars are available from the maven

**Comment by Riernar** [comment]
@KnightMiner 

Am I matching heat recipe, mix recipes and additives removals correctly ? I'd like that you don't have to specify the exact amount in zenscript for the removal to work.

**Comment by KnightMiner** [comment]
I have not had a chance to look into it too in-depth, but I think it looks right for the fluids. There's a better way to handle the item ingredient matching I think, I'll leave some comments on that tonight when I have some free time

**Comment by Riernar** [comment]
@KnightMiner 
I'd like your input on that, especially for the `MixRecipeManager`. I get a `RecipeMatch` from storing the user's instructions and one from the event, and I don't know how to test if the user's one includes the event's one - previous code throws `CannotCast` as the `List<ItemStack>` cannot be casted to `NonNullList<ItemStack>`.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: This could be IIngredient, especially since most fuels are ore dictionary

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: I assume the issue with FluidStack to FluidStack was duplicate fluids deleted some combinations. I would suggest simplifying this by using `List<Pair<FluidStack,FluidStack>>` as that gets rid of the duplicate key problem. Just make sure you use the right pair so it exists on both client and server, the Inspirations cauldron class should have an example.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeBuilder.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeBuilder.java]: Probably don't need this comment, its the same as the above line

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java]: Don't really need to use RecipeMatch for the removals since it never gets passed to the recipe, just use IIngredient directly for the match

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java]: This is not going to work well, `getInputs()` gets you a list of every possible match for things such as oredict, esentially giving you the list shown in JEI. The matches method on the other hand is expecting a list of slots, which in the case of the high oven is always one slot so the list should be size 1. What you really want here is iterate the list of additives in the recipe and see if any match the ingredient

That would look something like
```
for (ItemStack stack : recipe.getAdditive().getInputs())
  if (removal.matches(stack)) {
    event.setCanceled(true);
    break;
  }
```
Of course converting `stack` into a NonNullList.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: I would suggest giving these different names just to make it easier on stacktraces

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [inline src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: After switching this to use IIngredient instead of IItemStack, I'd flip this to `IIngredient.matches(any from recipe)`

**Review by KnightMiner — CHANGES_REQUESTED** [review]
Looks like just a few more things that I would change then I think it will be good to go

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Not really. This is to handle multiple inputs yielding the same output, if such behaviour is added by addons or other mods.
In case the user wants to delete all recipes for the given output, no matter the input, the input set is set to `null` and the method won't bother iterating.
If only some specific input should be disabled, the set is defined, inputs are stored and compared against during event handling.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java]: Thanks ! I really misunderstood what `RecipeMatch.matches()` does. This should work properly now.

**Comment by Riernar** [comment]
I think this PR is ready for merging. Everything seems to work properly according to the test script and all use case we could think of have been covered.

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (KnightMiner) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Yes, but there really is no need to do it as a set, it does not matter if I cancel on input A, output then 5 entries later cancel on input B, output; just check if any pair in the list matches and return after the first match. Plus the odds of a user removing more than 1 input without wanting wildcard is pretty low, since the pair method already handles it no reason to make it less efficient in most cases for negligible benefit in one. On a related note, pretty sure you will get a NPE in the CT method if you first use null input then a specific one, and even if not I'm not even certain fluidstacks support being a hashmap key

Additionally, the map misses the case of null output and defined input as maps don't allow null as a key last I checked. Just make it a list of pairs (or simpleentry) and it handles the problem you are trying to solve with far simplier code.

**Comment by KnightMiner** [comment]
Test script needs a manageMixRexipe test since you added that method after the script. There is a place in the CT source code for you to commit they script for later by the way, so it's not just in this PR

**Comment by Riernar** [comment]
I did add a test script testing all methods, see [my latest commit](https://github.com/jaredlll08/ModTweaker/pull/728/commits/3037e1acadf9ece477b3d56cf74e1987d08c8b59)

I think I haven't forgot anything in it

File: src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java
Comment (Riernar) [reply src/main/java/com/blamejared/compat/tcomplement/highoven/HighOven.java]: Oh, you're right ! I changed that in the latest commit

**Comment by Riernar** [comment]
@KnightMiner I added your last suggestion (pairs).
Do you see anything else amiss I should change, or are we ready for merging ?

**Comment by KnightMiner** [comment]
Ah, did not notice the committed one, it did not show up on the changed files yesterday so I missed it.

~~I suggest you change one of the two manageMixRecipe calls to use a recipe that exists by default such as pig iron just so we know it works on base recipes. Doubt it would break there, but good to test anyways.~~ Edit: nevermind, I see the pig iron manage there.

Otherwise, I think it looks good. I still want to test the latest stuff in game before I give my full approval, but as far as code goes it looks good

**Comment by KnightMiner** [comment]
Just tested the changes in game and they looked good overall. I did discover a bug in Tinkers Complement that prevents mix recipes from working if there are fewer than 3 additives. I fixed it in build 40, but this will still work with the latest build Tinkers Complement without updating the build dependency.

**Review by KnightMiner — APPROVED** [review]
Just tested the changes in game and they looked good overall. I did discover a bug in Tinkers Complement that prevents mix recipes from working if there are fewer than 3 additives. I fixed it in build 40, but this will still work with the latest build Tinkers Complement without updating the build dependency.

**Comment by Riernar** [comment]
Maybe it didn't show up because it's a new file ?
Anyway, I'm glad everything works properly. I imagine the bug in TComplement's is due to internal code and doesn't affect the API. I must admit I did not check if the recipes actually worked and just trusted JEI ^^

Thanks a lot for your extensive help and advices on that compatibility, it's much better thanks to you ! Plus it's quite awesome to work with a mod author !

**Comment by Riernar** [comment]
Hi @jaredlll08 
We are ready for merging this support in ModTweaker !
I left the version handling to you, so you can keep everything consistent. There might be a slight merge conflict with #729 since I copy-pasted the mantle helper. The file are the same, I hope it's not too troublesome.
If you need anything else for that PR, feel free to ask !

**Comment by Riernar** [comment]
@KnightMiner I guess I forgot High Oven's melting overrides as requested in #541 , those seems to be supported by `TCompRegistry`.
I'll add a method in a bit

**Comment by Riernar** [comment]
@KnightMiner I added support for High Oven melting overrides as in #541 , this was pretty much a copy-past from the `Override` class.

@jaredlll08 Everything should be good now

**Comment by Riernar** [comment]
Hi @jaredlll08 ,
Do you know when you'll have time to review and eventually merge ? I know it can be pretty tedious, so if I can help you in any way, please ask !

**Comment by jaredlll08** [comment]
No idea. I am currently dealing with university projects, when that has calmed down, I will go through and look at all of modtweaker related stuff.

(even if I merged it now, it would still be a bit before a release)

**Comment by Riernar** [comment]
No problem, I'll just keep an eye to be around when/if you need me. Merge/release the way you think is best - no need for hasty merges while you're dealing with more important stuff.
Good luck with your projects !

(Btw, I won't be much available in August, if that matters)

**Comment by hoklai1997** [comment]
@jaredlll08 , do you think you'll have the chance to merge this soon? I was hoping to use the integration in a modpack, but if you don't think you'll get to merging the tcomplement support, then I can look at other options. Don't mean to rush you, just curious about a timeline, Thanks!

**Comment by kindlich** [comment]
Hey @Riernar 

short question since I only now realized that I merged the wiki PR already lol:
Since the original PR was filed you have added some more commits to this PR.
Does this affect the ZS classes, as in, is the wiki page you wrote back then still up to date with the commits you have added later?

**Comment by Riernar** [comment]
Hi @kindlich ,
Just had a look to make sure, I think the documentation is up to date. The last feature I added was High Oven Overrides' modifications, and the associated commit was pushed on the PR before the merge.
The md formatting is broken here and there, though, and some sentences are not very clear. I guess I should fix that at some point.

**Comment by Riernar** [comment]
Hi @hoklai1997 ,
Since ModTweaker is released under the MIT license, I *think* you can use and distribute a jar built from my fork, in the mean time (don't quote me on that though, I'm no legal expert). That's just a few gradle commands, if you need help feel free to ask.

If you do, please report any bug in the high oven compatibility, so I can fix them. Testing in an actual modpack might reveal some corner cases. And of course, don't forget to credit Jared ;)

**Comment by hoklai1997** [comment]
Thanks, sounds like things might be moving, so I think I’ll wait for the official release. However, I do have plenty of scripts using your compatibility, so I’ll definitely let you know if I find anything!

**Comment by Fr4gtastic** [comment]
Hello @jaredlll08,
Not trying to rush you or anything, but can we expect this PR to be merged anytime soon? If not, could the docs entry at least be updated to mention the fact that this feature is not yet implemented? It's a bit confusing at the moment. 
Thanks!

**Comment by hoklai1997** [comment]
Hey @Riernar, I did decide to build the jar for now, at least so I can continue development, I'll let you know if I run into anything! I'm hoping @jaredlll08 will get the chance to merge it officially before my release, but if not I'll checkin again.

**Pull request merged**
